File: crates/app/src/commands/bookmarks.rs
use std::sync::Mutex;

use mejiro_core::{
    bookmarks::{Bookmarks, NestedBookmarks},
    data::{FolderData, ToolbarBookmarkData},
};

use crate::error::AppError;

#[tauri::command]
pub fn get_nested_json(
    state: tauri::State<'_, Mutex<Bookmarks>>,
    index: usize,
) -> Result<NestedBookmarks, AppError> {
    if index == 0 {
        return Err(AppError::Other("index should not be 0".to_string()));
    }
    let bookmarks = state
        .lock()
        .map_err(|_| AppError::Mutex("can't get bookmarks".to_string()))?;
    Ok(bookmarks.to_nested(index)?)
}

#[tauri::command]
pub fn get_root_and_children_folders(
    state: tauri::State<'_, Mutex<Bookmarks>>,
) -> Result<Vec<FolderData>, AppError> {
    let bookmarks = state
        .lock()
        .map_err(|_| AppError::Mutex("can't get bookmarks".to_string()))?;
    Ok(bookmarks.get_root_and_children_folders()?)
}

#[tauri::command]
pub fn get_toolbar_bookmarks(
    state: tauri::State<'_, Mutex<Bookmarks>>,
) -> Result<Vec<ToolbarBookmarkData>, AppError> {
    let bookmarks = state
        .lock()
        .map_err(|_| AppError::Mutex("can't get bookmarks".to_string()))?;
    Ok(bookmarks.get_toolbar_bookmarks())
}

#[tauri::command]
pub fn add_bookmark(
    state: tauri::State<'_, Mutex<Bookmarks>>,
    url: String,
    title: Option<String>,
    top_level_index: usize,
) -> Result<String, AppError> {
    let mut bookmarks = state
        .lock()
        .map_err(|_| AppError::Mutex("can't get bookmarks".to_string()))?;
    bookmarks.add_bookmark(url, title, top_level_index)?;

    Ok(bookmarks.to_nested_json(top_level_index)?)
}

#[tauri::command]
pub fn append_bookmark_to_toolbar(
    state: tauri::State<'_, Mutex<Bookmarks>>,
    title: String,
    url: String,
    top_level_index: usize,
) -> Result<String, AppError> {
    let mut bookmarks = state
        .lock()
        .map_err(|_| AppError::Mutex("can't get bookmarks".to_string()))?;
    bookmarks.append_bookmark_to_toolbar(&title, &url)?;

    Ok(bookmarks.to_nested_json(top_level_index)?)
}

#[tauri::command]
pub fn remove_bookmark(
    state: tauri::State<'_, Mutex<Bookmarks>>,
    index: usize,
    top_level_index: usize,
) -> Result<String, AppError> {
    let mut bookmarks = state
        .lock()
        .map_err(|_| AppError::Mutex("can't get bookmarks".to_string()))?;
    bookmarks.remove_subtree(index)?;

    Ok(bookmarks.to_nested_json(top_level_index)?)
}

#[tauri::command]
pub fn update_bookmark_title(
    state: tauri::State<'_, Mutex<Bookmarks>>,
    index: usize,
    title: String,
    top_level_index: usize,
) -> Result<String, AppError> {
    let mut bookmarks = state
        .lock()
        .map_err(|_| AppError::Mutex("can't get bookmarks".to_string()))?;
    bookmarks.update_title(index, title)?;

    Ok(bookmarks.to_nested_json(top_level_index)?)
}

#[tauri::command]
pub fn add_folder(
    state: tauri::State<'_, Mutex<Bookmarks>>,
    parent_index: usize,
    title: String,
    top_level_index: usize,
) -> Result<String, AppError> {
    let mut bookmarks = state
        .lock()
        .map_err(|_| AppError::Mutex("can't get bookmarks".to_string()))?;
    bookmarks.add_folder(parent_index, &title)?;

    Ok(bookmarks.to_nested_json(top_level_index)?)
}

#[tauri::command]
pub fn insert_after(
    state: tauri::State<'_, Mutex<Bookmarks>>,
    source_index: usize,
    destination_index: usize,
    top_level_index: usize,
) -> Result<String, AppError> {
    let mut bookmarks = state
        .lock()
        .map_err(|_| AppError::Mutex("can't get bookmarks".to_string()))?;
    bookmarks.insert_after(source_index, destination_index)?;

    Ok(bookmarks.to_nested_json(top_level_index)?)
}

#[tauri::command]
pub fn insert_before(
    state: tauri::State<'_, Mutex<Bookmarks>>,
    source_index: usize,
    destination_index: usize,
    top_level_index: usize,
) -> Result<String, AppError> {
    let mut bookmarks = state
        .lock()
        .map_err(|_| AppError::Mutex("can't get bookmarks".to_string()))?;
    bookmarks.insert_before(source_index, destination_index)?;

    Ok(bookmarks.to_nested_json(top_level_index)?)
}

#[tauri::command]
pub fn append_to_child(
    state: tauri::State<'_, Mutex<Bookmarks>>,
    source_index: usize,
    destination_index: usize,
    top_level_index: usize,
) -> Result<String, AppError> {
    let mut bookmarks = state
        .lock()
        .map_err(|_| AppError::Mutex("can't get bookmarks".to_string()))?;
    bookmarks.append_to_child(source_index, destination_index)?;

    Ok(bookmarks.to_nested_json(top_level_index)?)
}

#[tauri::command]
pub fn prepend_to_child(
    state: tauri::State<'_, Mutex<Bookmarks>>,
    source_index: usize,
    destination_index: usize,
    top_level_index: usize,
) -> Result<String, AppError> {
    let mut bookmarks = state
        .lock()
        .map_err(|_| AppError::Mutex("can't get bookmarks".to_string()))?;
    bookmarks.prepend_to_child(source_index, destination_index)?;

    Ok(bookmarks.to_nested_json(top_level_index)?)
}

#[tauri::command]
pub fn set_is_open(
    state: tauri::State<'_, Mutex<Bookmarks>>,
    index: usize,
    is_open: bool,
    top_level_index: usize,
) -> Result<String, AppError> {
    let mut bookmarks = state
        .lock()
        .map_err(|_| AppError::Mutex("can't get bookmarks".to_string()))?;
    bookmarks.set_is_open(index, is_open)?;

    Ok(bookmarks.to_nested_json(top_level_index)?)
}

#[tauri::command]
pub fn toggle_is_open(
    state: tauri::State<'_, Mutex<Bookmarks>>,
    index: usize,
    top_level_index: usize,
) -> Result<String, AppError> {
    let mut bookmarks = state
        .lock()
        .map_err(|_| AppError::Mutex("can't get bookmarks".to_string()))?;
    bookmarks.toggle_is_open(index)?;

    Ok(bookmarks.to_nested_json(top_level_index)?)
}

File: apps/desktop/src/stores/bookmarks.ts
import { createWithSignal } from "solid-zustand";

import { Invoke } from "../invokes";
import type { NestedBookmark, ToolbarBookmarkData } from "../types";

export const getFolders = async () => {
  const folders = await Invoke.GetRootAndChildrenFolders();
  return folders;
};

type BookmarkState = {
  bookmarks: NestedBookmark | null;
  toolbarBookmarks: ToolbarBookmarkData[];
  isTreeLocked: boolean;
  activeIndex: number | null;
  editingIndex: number | null;
  setTreeLockState: (value: boolean) => void;
  setActiveIndex: (index: number | null) => void;
  setEditingIndex: (index: number | null) => void;
  getCurrentTopLevel: () => number;
  getBookmarks: (index: number) => Promise<void>;
  getToolbarBookmarks: () => Promise<void>;
  addBookmark: (title: string, url: string) => Promise<void>;
  appendBookmarkToToolbar: (title: string, url: string) => Promise<void>;
  removeBookmark: (index: number) => Promise<void>;
  updateBookmarkTitle: (index: number, title: string) => Promise<void>;
  addFolder: (parentIndex: number, title: string) => Promise<void>;
  insertAfter: (sourceIndex: number, destinationIndex: number) => Promise<void>;
  insertBefore: (sourceIndex: number, destinationIndex: number) => Promise<void>;
  appendToChild: (sourceIndex: number, destinationIndex: number) => Promise<void>;
  prependToChild: (sourceIndex: number, destinationIndex: number) => Promise<void>;
  setIsOpen(index: number, isOpen: boolean): void;
  toggleIsOpen: (index: number) => Promise<void>;
};

export const useBookmarkState = createWithSignal<BookmarkState>((set, get) => ({
  bookmarks: null,
  toolbarBookmarks: [],
  isTreeLocked: false,
  activeIndex: null,
  editingIndex: null,
  setTreeLockState: (value) => {
    set(() => ({ isTreeLocked: value }));
  },
  setActiveIndex: (index) => {
    set(() => ({ activeIndex: index }));
  },
  setEditingIndex: (index) => {
    set(() => ({ editingIndex: index }));
  },
  getCurrentTopLevel: () => {
    const bookmarks = get().bookmarks;
    if (bookmarks !== null) {
      return bookmarks.index;
    }
    return 1;
  },
  getToolbarBookmarks: async () => {
    const toolbarBookmarks = await Invoke.GetToolbarBookmarks();
    set(() => ({ toolbarBookmarks }));
  },
  getBookmarks: async (index) => {
    // can't accept index 0 because indextree starts from 1
    if (index >= 1) {
      // SetIsOpen returns the updated bookmarks that open top level folder
      const topLevelIndex = index;
      const bookmarks = await Invoke.SetIsOpen(index, true, topLevelIndex);
      set(() => ({ bookmarks }));
    }
  },
  addBookmark: async (title, url) => {
    const topLevelIndex = get().getCurrentTopLevel();
    const res = await Invoke.AddBookmark(title, url, topLevelIndex);
    set(() => ({ bookmarks: res.bookmarks }));
  },
  appendBookmarkToToolbar: async (title, url) => {
    const topLevelIndex = get().getCurrentTopLevel();
    const bookmarks = await Invoke.AppendBookmarkToToolbar(title, url, topLevelIndex);
    set(() => ({ bookmarks }));
  },
  removeBookmark: async (index) => {
    const topLevelIndex = get().getCurrentTopLevel();
    const bookmarks = await Invoke.RemoveBookmark(index, topLevelIndex);
    set(() => ({ bookmarks }));
  },
  updateBookmarkTitle: async (index, title) => {
    const topLevelIndex = get().getCurrentTopLevel();
    const bookmarks = await Invoke.UpdateBookmarkTitle(index, title, topLevelIndex);
    set(() => ({ bookmarks }));
  },
  addFolder: async (parentIndex, title) => {
    const topLevelIndex = get().getCurrentTopLevel();
    const res = await Invoke.AddFolder(parentIndex, title, topLevelIndex);
    set(() => ({ bookmarks: res.bookmarks, editingIndex: res.index }));
  },
  insertAfter: async (sourceIndex, destinationIndex) => {
    const topLevelIndex = get().getCurrentTopLevel();
    const bookmarks = await Invoke.InsertAfter(sourceIndex, destinationIndex, topLevelIndex);
    set(() => ({ bookmarks }));
  },
  insertBefore: async (sourceIndex, destinationIndex) => {
    const topLevelIndex = get().getCurrentTopLevel();
    const bookmarks = await Invoke.InsertBefore(sourceIndex, destinationIndex, topLevelIndex);
    set(() => ({ bookmarks }));
  },
  appendToChild: async (sourceIndex, destinationIndex) => {
    const topLevelIndex = get().getCurrentTopLevel();
    const bookmarks = await Invoke.AppendToChild(sourceIndex, destinationIndex, topLevelIndex);
    set(() => ({ bookmarks }));
  },
  prependToChild: async (sourceIndex, destinationIndex) => {
    const topLevelIndex = get().getCurrentTopLevel();
    const bookmarks = await Invoke.PrependToChild(sourceIndex, destinationIndex, topLevelIndex);
    set(() => ({ bookmarks }));
  },
  setIsOpen: async (index, isOpen) => {
    const topLevelIndex = get().getCurrentTopLevel();
    const bookmarks = await Invoke.SetIsOpen(index, isOpen, topLevelIndex);
    set(() => ({ bookmarks }));
  },
  toggleIsOpen: async (index) => {
    const topLevelIndex = get().getCurrentTopLevel();
    const bookmarks = await Invoke.ToggleIsOpen(index, topLevelIndex);
    set(() => ({ bookmarks }));
  },
}));
